pallets/baby-liminal/src/lib.rs

```diff
@@ -114,40 +154,61 @@ pub mod pallet {
         #[pallet::call_index(0)]
         #[pallet::weight(T::WeightInfo::store_key(key.len() as u32))]
         pub fn store_key(
-            _origin: OriginFor<T>,
+            origin: OriginFor<T>,
             identifier: VerificationKeyIdentifier,
             key: Vec<u8>,
         ) -> DispatchResult {
-            Self::bare_store_key(identifier, key).map_err(|e| e.into())
+            Self::bare_store_key(origin, identifier, key).map_err(|e| e.into())
         }
 
         /// Deletes a key stored under `identifier` in `VerificationKeys` map.
         ///
-        /// Can only be called by a root account.
+        /// Returns the deposit locked. Can only be called by the key owner.
         #[pallet::call_index(1)]
-        #[pallet::weight(T::DbWeight::get().writes(1))]
+        #[pallet::weight(T::DbWeight::get().reads(2) + T::DbWeight::get().writes(2))]
         pub fn delete_key(
             origin: OriginFor<T>,
             identifier: VerificationKeyIdentifier,
         ) -> DispatchResult {
-            ensure_root(origin)?;
+            let who = ensure_signed(origin).map_err(|_| Error::<T>::BadOrigin)?;
+            let owner = VerificationKeyOwners::<T>::get(identifier)
+                .ok_or(Error::<T>::UnknownVerificationKeyIdentifier)?;
+
+            ensure!(who == owner, Error::<T>::NotOwner);
+
+            let deposit = VerificationKeyDeposits::<T>::take((&owner, &identifier)).unwrap(); // cannot fail since the key has owner and owner must have made a deposit
+            T::Currency::unreserve(&owner, deposit);
+
             VerificationKeys::<T>::remove(identifier);
-            Self::deposit_event(Event::VerificationKeyDeleted);
+            Self::deposit_event(Event::VerificationKeyDeleted(who, identifier));
             Ok(())
         }
 
         /// Overwrites a key stored under `identifier` in `VerificationKeys` map with a new value `key`
         ///
         /// Fails if `key.len()` is greater than `MaximumVerificationKeyLength`.
-        /// Can only be called by a root account.
+        /// Can only be called by the original owner of the key.
         #[pallet::call_index(2)]
         #[pallet::weight(T::WeightInfo::overwrite_key(key.len() as u32))]
         pub fn overwrite_key(
             origin: OriginFor<T>,
             identifier: VerificationKeyIdentifier,
             key: Vec<u8>,
         ) -> DispatchResult {
-            ensure_root(origin)?;
+            let who = ensure_signed(origin).map_err(|_| Error::<T>::BadOrigin)?;
+            let owner = VerificationKeyOwners::<T>::get(identifier);
+            let previous_key = VerificationKeys::<T>::get(identifier);
+
+            if let Some(owner) = owner {
```

Comment:
When would that be a `None`? If someone calles `ovewrite_key` instead of `store_key`?

Reply:
yes

Reply:
Then why not, instead, remove the `overwrite_key` method and just tell people to use `insert_key` if they want to overwrite?

Reply:
Because you cannot store a key with the same identifier twice

Reply:
Even if we tell people, they can still use this call and because of that we need this check.

Reply:
Not if we didn't have `overwrite_key` at all and do everything in `store_key`: detect if there already exists an entry for that key and if so overwrite it.

Reply:
Generally speaking there is a couple good reasons to keep these separate - one is the api of the StorageMap, another is that it gets hard to benchmark two paths like that

Reply:
OK. That's a good point, I can understand that. But aren't you "violating" your own rule right now in `overwrite_key`? If it's user is calling `overwrite_*` instead of `store_*` then you are using the method for not something it was intended to -- it should also be hard to benchmark.

Reply:
The answer is provenance and likeness - it's just easier to support insert for overwrite, it incurs no additional db operations and these dominate the weights.